### PR TITLE
DEVPROD-8873 Pass a github app token to fetch when fetching task data for spawn hosts

### PR DESCRIPTION
**Pull request by malikchaya2**
DEVPROD-8873

### Description
When spawning a host from a task, we used to be able to rely on the ssh key on the hosts to clone the repos and modules when fetching data associated with a task. We won't be able to rely on that once there is no longer an ssh key. This passes a app token to fetch to use instead.  

### Testing
Tested on staging by creating a [task](evg_DEVPROD_5742_ubuntu2204_small_just_file_stuff_patch_9efddbb7e5fa5ffed5671a60562f00f21335eec9_669fc8c9713dd40007d61bcf_24_07_23_15_14_50) with two modules (one private) and spawning a [host](https://spruce-staging.corp.mongodb.com/host/i-06c42fcf8f802f2c7) on a distro that doesn't have an ssh key.

### CODE REVIEW

File: graphql/util.go [hunk 1]
```diff
@@ -909,6 +919,53 @@ func getHostRequestOptions(ctx context.Context, usr *user.DBUser, spawnHostInput
...
+	project, err := model.TranslateProject(pp)
+	if err != nil || project == nil {
+		return nil, nil, errors.Wrap(err, "translating project")
+	}
+	variant := project.FindBuildVariant(t.BuildVariant)
+	if variant == nil {
+		return nil, nil, errors.Wrapf(err, "finding build variant '%s' in config", t.BuildVariant)
+	}
+
+	var manifestModules []manifest.Module
```
Comment (ablack12) [inline graphql/util.go:950]: Why are we using the manifest modules instead of the project modules? Then we wouldn't need any of this name stuff.

File: model/project.go [hunk 2]
```diff
@@ -473,6 +473,13 @@ func (m Module) GetOwnerAndRepo() (string, string, error) {
 	return m.Owner, m.Repo, nil
 }
 
+// CreateName creates a name based on owner, repo and branch
+func (m Module) CreateName() string {
```
Comment (ablack12) [inline model/project.go:477]: I think this isn't great for module/project.go because this struct already has a name field.

File: operations/fetch.go [hunk 3]
```diff
@@ -58,6 +60,14 @@ func Fetch() cli.Command {
 				Name:  joinFlagNames(tokenFlagName, "k"),
 				Usage: "GitHub API token",
 			},
+			cli.BoolFlag{
+				Name:  useAppTokenName,
+				Usage: "when using a token, use this to indicate that the token is for a github app and not an oauth token",
```
Comment (ablack12) [inline operations/fetch.go:65]: nit: GitHub since this is help text

File: model/host/host.go [hunk 4]
```diff
@@ -315,6 +319,22 @@ type ProvisionOptions struct {
...
+	// ProjectOwner is the owner of the project associated with the task.
+	ProjectOwner string `bson:"project_owner" json:"project_owner"`
+	// Modules is a list of modules associated with the task.
+	Modules []manifest.Module `bson:"modules" json:"modules"`
+
+	// GithubAppToken is the token used to fetch task data from Github.
+	GithubAppToken string `bson:"github_app_token" json:"github_app_token"`
+	// ModuleTokens is a list of github tokens for each module associated with the task
+	// in the format 'moduleOwner_moduleRepo:token'
+	ModuleTokens []string `bson:"module_tokens" json:"module_tokens"`
```
[the remark this answers is not in the document]
Comment (ablack12) [reply model/host/host.go:335]: I see that this is the format we pass into the CLI, but could we format it at that time rather than storing it like this?

File: rest/model/host.go [hunk 5]
```diff
@@ -62,18 +63,21 @@ func (apiOpts *APIProvisionOptions) BuildFromService(opts host.ProvisionOptions)
...
-	UserData              string     `json:"userdata" yaml:"userdata_file"`
-	SetupScript           string     `json:"setup_script" yaml:"setup_file"`
-	UseProjectSetupScript bool       `json:"use_setup_script_path" yaml:"use_setup_script_path"`
-	Tag                   string     `yaml:"tag"`
-	InstanceTags          []host.Tag `json:"instance_tags" yaml:"instance_tags"`
-	InstanceType          string     `json:"instance_type" yaml:"type"`
-	NoExpiration          bool       `json:"no_expiration" yaml:"no-expire"`
+	DistroID              string            `json:"distro" yaml:"distro"`
+	TaskID                string            `json:"task" yaml:"task"`
+	ProjectRepo           string            `bson:"project_repo" json:"project_repo"`
```
Comment (ablack12) [inline rest/model/host.go:68]: Do we need these if we have the task ID? (If we do, can we move them to the bottom with a comment about why we're adding them?)

File: units/host_setup_script.go [hunk 6]
```diff
@@ -116,6 +118,21 @@ func (j *hostSetupScriptJob) Run(ctx context.Context) {
 			j.AddRetryableError(errors.Wrap(err, "checking if task data is fetched yet"))
 			return
 		}
+		// once task data is fetched, we no longer need the github token. Revoke it here
```
Comment (ablack12) [inline units/host_setup_script.go:121]: nit: capitalize "once" since this is a full sentence

File: units/host_setup_script.go [hunk 7]
```diff
@@ -116,6 +118,21 @@ func (j *hostSetupScriptJob) Run(ctx context.Context) {
...
+		// in case something went wrong and it's still around.
+		if f := j.host.ProvisionOptions.FetchOpts; f != nil {
+			_ = thirdparty.RevokeInstallationToken(ctx, f.GithubAppToken)
```
Comment (ablack12) [inline units/host_setup_script.go:124]: GetInstallationToken uses a cached token, so we shouldn't revoke I think. cc @Kimchelly can you check this?

**Review by ablack12 — CHANGES_REQUESTED** [review]

File: units/host_setup_script.go (hunk 7 above)
Comment (Kimchelly) [reply units/host_setup_script.go:124]: That's correct, as per the doc comment on `RevokeInstallationToken`, it's invalid to call it with a cached token. It's also the case that this will only revoke if the user chooses to run a custom setup script (which is pretty uncommon). If they don't set a custom setup script, this job never runs, so it'll never revoke the token.

If this is being given to a user's spawn host, it probably shouldn't use a cached token. The token caching is for make internal Evergreen usage more efficient, not for users.

Comment (malikchaya2) [reply units/host_setup_script.go:124]: This PR is from two weeks ago so it predates the work to start caching Installation tokens. I'll read it again to make sure it still works as expected. 
This is using a more scoped down read only token, so I don't think it makes sense to use the cache, as those are full-permission tokens.

**Comment by ablack12** [comment]
@Kimchelly would you mind taking over this PR since I'll be offline? Most open questions relate to your new work.

File: model/host/hostutil.go [hunk 8]
```diff
@@ -1038,6 +1046,74 @@ func (h *Host) SpawnHostSetupCommands(settings *evergreen.Settings) (string, err
...
+// PopulateGithubToken adds a read-only token for owner/repo associated with the task
+// to h.ProvisionOptions
+func (h *Host) PopulateGithubToken(ctx context.Context) error {
+	p := h.ProvisionOptions
+	if p.TaskId == "" {
+		return errors.New("missing task id")
+	}
+	if p.FetchOpts == nil || p.FetchOpts.ProjectOwner == "" || p.FetchOpts.ProjectRepo == "" {
+		grip.Warning(message.Fields{
+			"message": "missing project owner or repo when populating token for spawn host",
```
[the remark this answers is not in the document]
Comment (malikchaya2) [reply model/host/hostutil.go:1061]: I mostly added it because there isn't much visibility into fetch failing as it runs in memory on the Jasper service, which doesn't record stdout/stderr logs. So the guidance is usually to just run fetch again. I was trying to add logs where I can. 
But I see your point. I'll remove them.

File: units/host_setup_script.go [hunk 9]
```diff
@@ -116,6 +117,14 @@ func (j *hostSetupScriptJob) Run(ctx context.Context) {
 			j.AddRetryableError(errors.Wrap(err, "checking if task data is fetched yet"))
 			return
 		}
+		// Once task data is fetched, we no longer need the github token. Revoke it here
+		// in case something went wrong and it's still around.
+		if f := j.host.ProvisionOptions.FetchOpts; f != nil {
```
[the remark this answers is not in the document]
Comment (malikchaya2) [reply units/host_setup_script.go:122]: I accidentally removed that line in the refactor. I'll add it back in.

[the remark this answers is not in the document]
Comment (malikchaya2) [reply units/host_setup_script.go]: done

File: thirdparty/github.go [hunk 10]
```diff
@@ -336,15 +340,33 @@ func getGithubClient(token, caller string, config retryConfig) *evergreen.GitHub
...
+}
+
+// GetInstallationToken creates an installation token using Github app auth.
+// It will not try to use the cache. This should only be used if there is a reason why a
+// cached token should not be used. If creating a token fails it will return the legacyToken.
+func GetInstallationToken(ctx context.Context, owner, repo string, opts *github.InstallationTokenOptions) (string, error) {
+	return installationToken(ctx, owner, repo, opts, false)
+}
+
+func installationToken(ctx context.Context, owner, repo string, opts *github.InstallationTokenOptions, useCached bool) (string, error) {
```
[the remark this answers is not in the document]
Comment (malikchaya2) [reply thirdparty/github.go:358]: I'm okay with either setup. There's also an argument to be made that it should always default to the cached token unless there is a specific reason why a non cached token is needed, which I believe will be rare. I can also solve this by adding more comments. 
@Kimchelly I'll leave it up to you.

[the remark this answers is not in the document]
Comment (Kimchelly) [reply thirdparty/github.go:358]: I believe the decision to use a cached or uncached token has to be made deliberately. Caching is meant to be a workaround to the problem where some of Evergreen's most common operations (e.g. getting GitHub API info, or getting a token for git.get_project cloning)  where we know that users can't directly interact with the token except in a specific/highly controlled way. If the operation rarely happens or somehow a user could directly get the token, then it's better to give them a fresh (no cache) token because Evergreen needs to reuse that token for its own internal purposes.

Since fetching artifact data is a pretty infrequent operation (I'd be surprised if more than a couple of hosts per day are spawned with task data) and users could get the token by querying Jasper, I'd say that `evergreen fetch` should not use a cached token at all.

Comment (Kimchelly) [inline thirdparty/github.go:358]: Re the "use a boolean" thing, I'd actually argue that this setup of using a boolean and making two wrapper functions is basically the exact same as calling one function vs. the other. The methods above already re-implement the `GetInstallationToken` and `GetCachedInstallationToken` methods, so I'm not sure they're really much better than the existing interface. I'd argue that unless we force all callers to use this method (and remove the other two `GetInstallationToken` and `getCachedInstallationToken` functions), this is just duplicating the existing methods from the top-level package.

File: model/host/hostutil.go [hunk 11]
```diff
@@ -1038,6 +1046,83 @@ func (h *Host) SpawnHostSetupCommands(settings *evergreen.Settings) (string, err
...
+		return errors.New("missing project owner or project repo")
+	}
+
+	opts := &github.InstallationTokenOptions{
+		Permissions: &github.InstallationPermissions{
+			Contents: utility.ToStringPtr(thirdparty.GithubPermissionRead),
+		},
+	}
+
+	token, err := thirdparty.GetInstallationToken(ctx, p.FetchOpts.ProjectOwner, p.FetchOpts.ProjectRepo, opts)
```
Comment (Kimchelly) [inline model/host/hostutil.go:1069]: nit: it seems like the new thirdparty functions are just wrapping the exact same methods that already exist in the `evergreen` package. I'd advocate strongly for not duplicating functionality into a different package, and just calling the `evergreen` package functions directly.

File: units/provisioning_setup_host.go [hunk 12]
```diff
@@ -502,6 +506,7 @@ func (j *setupHostJob) provisionHost(ctx context.Context, settings *evergreen.Se
 				"host_id":   j.host.Id,
 				"job":       j.ID(),
 			}))
+			j.host.RevokeGithubTokens(ctx)
```
Comment (Kimchelly) [inline units/provisioning_setup_host.go:509]: Did we test that this revokation works with `evergreen fetch`? It should be invalid to revoke the token here because the `evergreen fetch` runs in the background on the host, meaning it can revoke while `evergreen fetch` is still running.

If we really want to revoke the tokens safely, we could add an option to `evergreen fetch` to auto-revoke the tokens when the command is finished. That way, there's no need for the app servers to try to figure out when the host has finished running the command.

File: agent/command/git.go [hunk 13]
```diff
@@ -735,6 +735,7 @@ func (c *gitFetchProject) fetchModuleSource(ctx context.Context,
 		appToken, err := comm.CreateInstallationToken(ctx, td, opts.owner, opts.repo)
 		if err == nil {
 			opts.token = appToken
+			opts.method = cloneMethodAccessToken
```
Comment (Kimchelly) [inline agent/command/git.go:738]: Is this relevant to the ticket? Seems kind of random since this is for git.get_project, not `evergreen fetch`.

File: cloud/ec2_fleet.go [hunk 14]
```diff
@@ -144,6 +144,7 @@ func (m *ec2FleetManager) SpawnHost(ctx context.Context, h *host.Host) (*host.Ho
 		"host_provider": h.Distro.Provider,
 		"distro":        h.Distro.Id,
 	})
+	h.RevokeGithubTokens(ctx)
```
Comment (Kimchelly) [inline cloud/ec2_fleet.go:147]: Why is the token revoked here? It seems like the token wouldn't even exist yet by this point because `PopulateGitHubTokens` hasn't been called yet.

File: model/host/host.go [hunk 15]
```diff
@@ -304,6 +304,9 @@ type ProvisionOptions struct {
 	// artifacts for the given task.
 	TaskId string `bson:"task_id" json:"task_id"`
 
+	//FetchOps contains data needed to fetch task data
```
Comment (Kimchelly) [inline model/host/host.go:307]: nit: `// FetchOpts`

File: model/host/host.go [hunk 16]
```diff
@@ -304,6 +304,9 @@ type ProvisionOptions struct {
...
+	FetchOpts *FetchOpts `bson:"fetch_options,omitempty" json:"fetch_options,omitempty"`
```
Comment (Kimchelly) [inline model/host/host.go:308]: It would simplify things a lot (both for this PR and for the future) if we did the thing I proposed in my other comment where we 1. revoke in the CLI command itself rather than on the app server side, and then 2. rather than initializing the fetch options when the host is inserted into the DB, move all the logic to figure out the owner/repo/modules/tokens into [this function](https://github.com/evergreen-ci/evergreen/blob/37d5980bdca8f361cec36529303df6da98ae9323/model/host/hostutil.go#L1104-L1118). By doing that, we could lazily populate the tokens right when the `evergreen fetch` command is being made and pass the tokens along to the command. The command will eventually revoke them itself. It also has the advantage of simplicity, because instead of having to pass around all this host state all over the place, we only use it in the one place where it's actually relevant (i.e. making the `evergreen fetch` command).

Doing it that way also has a big advantage of avoiding persisting GitHub tokens in plaintext in the DB. Since we're trying to move away from storing plaintext secrets in the DB, storing the tokens like this in the host document will eventually become tech debt that we have to figure out how to move out of the DB later on.

File: graphql/util.go [hunk 17]
```diff
@@ -909,6 +918,54 @@ func getHostRequestOptions(ctx context.Context, usr *user.DBUser, spawnHostInput
...
+	v, err := model.VersionFindOne(model.VersionById(t.Version))
+	if err != nil {
+		return nil, nil, errors.Wrap(err, "finding version")
+	}
+	pp, err := model.ParserProjectFindOneByID(ctx, settings, v.ProjectStorageMethod, v.Id)
+	if err != nil || pp == nil {
+		return nil, nil, errors.Wrap(err, "finding project ref")
+	}
+	// Then fetch each of the modules
+	project, err := model.TranslateProject(pp)
```
Comment (Kimchelly) [inline graphql/util.go:940]: Do we need to re-translate the entire project just to get the module info? This could be pretty expensive, so it would be good to avoid if possible?

Rather than re-translate the whole project, is it possible to just query the version manifest to figure out what modules are relevant to the task?

File: rest/model/host.go [hunk 18]
```diff
@@ -80,6 +80,13 @@ type HostRequestOptions struct {
 	HomeVolumeSize       int        `json:"home_volume_size" yaml:"home_volume_size"`
 	HomeVolumeID         string     `json:"home_volume_id" yaml:"home_volume_id"`
 	Expiration           *time.Time `json:"expiration" yaml:"expiration"`
+
+	// The project information needed to fetch task data. Fetching it when the host request
+	// options are built instead of using the task id to fetch it when the host is
+	// created saves a few db calls.
+	ProjectRepo    string               `bson:"project_repo" json:"project_repo"`
+	ProjectOwner   string               `bson:"project_owner" json:"project_owner"`
+	ProjectModules []host.ProjectModule `bson:"project_modules" json:"project_modules"`
```
Comment (Kimchelly) [inline rest/model/host.go:89]: I don't think we need to include these in the REST data model.

Also, if we're going to have these in the REST data model, it should be JSON and YAML tags, not BSON and JSON, since we never persist the REST model in the DB.

**Review by Kimchelly — CHANGES_REQUESTED** [review]

File: operations/fetch.go [hunk 19]
```diff
@@ -194,18 +223,25 @@ func fetchSource(ctx context.Context, ac, rc *legacyClient, comm client.Communic
...
-	token      string
-	depth      uint
+	owner        string
+	repository   string
+	revision     string
+	rootDir      string
+	branch       string
+	token        string
+	depth        uint
+	is_app_token bool
```
Comment (Kimchelly) [inline operations/fetch.go:233]: Per Go style, `isAppToken`

File: operations/fetch.go [hunk 20]
```diff
@@ -194,18 +223,25 @@ func fetchSource(ctx context.Context, ac, rc *legacyClient, comm client.Communic
...
 
 func clone(opts cloneOptions) error {
+	var cloneArgs []string
 	// clone the repo first
-	cloneArgs := []string{"clone", thirdparty.FormGitURL("github.com", opts.owner, opts.repository, opts.token)}
+	if opts.is_app_token {
+		cloneArgs = []string{"clone", thirdparty.FormGitURLForApp("github.com", opts.owner, opts.repository, opts.token)}
+	} else {
+		cloneArgs = []string{"clone", thirdparty.FormGitURL("github.com", opts.owner, opts.repository, opts.token)}
+	}
```
Comment (Kimchelly) [inline operations/fetch.go:243]: Rather than having to pass around more information in the CLI and in the agent about whether it's an OAuth token vs. an installation token, could we instead use the token's prefix to determine the clone URL? Feel like that would eliminate the need to pass around the token type. [This link](https://github.blog/engineering/platform-security/behind-githubs-new-authentication-token-formats/#identifiable-prefixes) suggests that they always use `ghp_` for personal access tokens and `ghs_` for installation tokens, so we could use that prefix to decide between `https://github.com:x-oauth-basic@...` and `https://github.com:x-access-token@`, rather than us having to manually pass in CLI options or figure out the token type in the agent.

Comment (malikchaya2) [reply operations/fetch.go:243]: I've shied away from that because as far as I know, they don't really provide a guarantee to not change this format, making it safer for the long run to keep track of it ourselves.

File: agent/command/git.go (hunk 13 above)
Comment (malikchaya2) [reply agent/command/git.go:738]: it's not, I just noticed it and it was just easier to roll it in here than create a separate PR.

File: units/provisioning_setup_host.go [hunk 21]
```diff
@@ -833,3 +839,53 @@ func parseLsblkOutput(lsblkOutput string) ([]blockDevice, error) {
...
+	var modules map[string]*manifest.Module
+	t, err := task.FindOneId(taskId)
+	if err != nil && t != nil {
+		mfest, err := manifest.FindFromVersion(t.Version, t.Project, t.Revision, t.Requester)
+		if err != nil {
+			modules = mfest.Modules
+		}
+		p, err := model.FindMergedProjectRef(t.Project, t.Version, false)
+		if err != nil {
+			projectOwner = p.Owner
```
Comment (Kimchelly) [inline units/provisioning_setup_host.go:860]: Need to nil check `p` before dereferencing since `model.FindMergedProjectRef` can return nil result if it can't find the manifest.

File: units/provisioning_setup_host.go [hunk 22]
```diff
@@ -833,3 +839,53 @@ func parseLsblkOutput(lsblkOutput string) ([]blockDevice, error) {
...
+	// because if the repo is not private, cloning the repo will still work.
+	// Either way, we should still spin up the host even if we can't
+	// fetch the data.
+	var projectOwner, projectRepo string
+	var modules map[string]*manifest.Module
+	t, err := task.FindOneId(taskId)
+	if err != nil && t != nil {
+		mfest, err := manifest.FindFromVersion(t.Version, t.Project, t.Revision, t.Requester)
+		if err != nil {
+			modules = mfest.Modules
```
Comment (Kimchelly) [inline units/provisioning_setup_host.go:856]: Need to nil check `mfest` since `manifest.FindFromVersion` can return nil result if it can't find the manifest.

File: units/provisioning_setup_host.go [hunk 23]
```diff
@@ -833,3 +839,53 @@ func parseLsblkOutput(lsblkOutput string) ([]blockDevice, error) {
 
 	return devices.BlockDevices, nil
 }
+
+// GetGithubTokensForTask returns a read-only token for owner/repo associated with the task
+// and a read-only token for each module associated with the task.
+func GetGithubTokensForTask(taskId string, ctx context.Context) (string, []string) {
```
Comment (Kimchelly) [inline units/provisioning_setup_host.go:845]: Can we add a unit test for this?

File: thirdparty/git.go [hunk 24]
```diff
@@ -271,6 +271,9 @@ func ParseGitUrl(url string) (string, string, error) {
 		url = strings.TrimPrefix(url, httpsPrefix)
 		split := strings.Split(url, "/")
 		if len(split) != 3 {
+			if split[0] == "api.github.com" && len(split) > 3 {
```
Comment (Kimchelly) [inline thirdparty/git.go:274]: Can we clarify that this is for the special case where we're using a GitHub API URL of the form `https://api.github.com/repos/<owner>/<repo>` rather than a git clone URL? Alternatively, since this function is meant to parse `git clone` URLs and the module URL is actually a GitHub API URL, consider making a separate function to parse GitHub API URLs.

File: rest/route/service.go [hunk 25]
```diff
@@ -265,6 +265,7 @@ func AttachHandler(app *gimlet.APIApp, opts HandlerOpts) {
 	app.AddRoute("/versions/{version_id}/builds").Version(2).Get().Wrap(requireUser, viewTasks).RouteHandler(makeGetVersionBuilds(env))
 	app.AddRoute("/versions/{version_id}/restart").Version(2).Post().Wrap(requireUser, editTasks).RouteHandler(makeRestartVersion())
 	app.AddRoute("/versions/{version_id}/annotations").Version(2).Get().Wrap(requireUser, viewAnnotations).RouteHandler(makeFetchAnnotationsByVersion())
+	app.AddRoute("/tasks/{task_id}/revoke_github_dynamic_access_tokens").Version(2).Delete().Wrap(viewTasks).RouteHandler(makeRevokeGitHubDynamicAccessTokens())
```
Comment (Kimchelly) [inline rest/route/service.go:268]: also nit: consider calling `/tasks/{task_id}/github_dynamic_access_tokens` since the agent route equivalent is called that. The `DELETE` part of the route signature should be clear enough that it's invalidating the token.

File: units/provisioning_setup_host.go [hunk 26]
```diff
@@ -833,3 +839,53 @@ func parseLsblkOutput(lsblkOutput string) ([]blockDevice, error) {
...
+		githubAppToken, _ = settings.CreateGitHubAppAuth().CreateInstallationToken(ctx, projectOwner, projectRepo, opts)
+
+	}
+
+	for moduleName, module := range modules {
+		if module.Repo == "" {
+			_, module.Repo, _ = thirdparty.ParseGitUrl(module.URL)
+		}
+		if module.Owner != "" && module.Repo != "" {
+			token, _ := settings.CreateGitHubAppAuth().CreateInstallationToken(ctx, module.Owner, module.Repo, opts)
```
Comment (Kimchelly) [inline units/provisioning_setup_host.go:885]: This can panic if `settings` wasn't retrieved.

File: units/provisioning_setup_host.go [hunk 27]
```diff
@@ -833,3 +839,53 @@ func parseLsblkOutput(lsblkOutput string) ([]blockDevice, error) {
...
+	if projectOwner != "" && projectRepo != "" && err == nil {
+		// Ignore any errors because if the repo is not private, cloning the repo will still work.
+		// Either way, we should still spin up the host even if we can't fetch the data.
+		githubAppToken, _ = settings.CreateGitHubAppAuth().CreateInstallationToken(ctx, projectOwner, projectRepo, opts)
+
+	}
+
+	for moduleName, module := range modules {
+		if module.Repo == "" {
+			_, module.Repo, _ = thirdparty.ParseGitUrl(module.URL)
```
Comment (Kimchelly) [inline units/provisioning_setup_host.go:882]: tiny nit: rather than setting the module's repo, would be a little cleaner to just define a local `var repo string` above, set it to `module.Repo`, and then overwrite it with the parsed result if `module.Repo` is empty.

**Review by Kimchelly — CHANGES_REQUESTED** [review]

File: units/provisioning_setup_host.go [hunk 28]
```diff
@@ -833,3 +839,53 @@ func parseLsblkOutput(lsblkOutput string) ([]blockDevice, error) {
 
 	return devices.BlockDevices, nil
 }
+
+// GetGithubTokensForTask returns a read-only token for owner/repo associated with the task
+// and a read-only token for each module associated with the task.
+func GetGithubTokensForTask(taskId string, ctx context.Context) (string, []string) {
+	// Do not error at any point while trying to populate github tokens
+	// because if the repo is not private, cloning the repo will still work.
```
[the remark this answers is not in the document]
Comment (malikchaya2) [reply units/provisioning_setup_host.go:847]: If the repo is not private, it's okay if for example, github is down and we couldn't get a token, because the clone will still work. If it's private and we failed to get a token for whatever reason, the clone will fail.

File: thirdparty/git.go (hunk 24 above)
Comment (malikchaya2) [reply thirdparty/git.go:274]: I think using the same function is fine because it's a generic "ParseGitUrl" function. I'll add a comment.

File: units/provisioning_setup_host.go (hunk 23 above)
Comment (malikchaya2) [reply units/provisioning_setup_host.go]: I don't want to generate actual tokens and I don't think it's worth the effort to mock it. I tested it thoroughly on staging.

File: rest/data/host.go [hunk 29]
```diff
@@ -104,7 +103,12 @@ func GenerateHostProvisioningScript(ctx context.Context, env evergreen.Environme
 			Message:    errors.Wrap(err, "generating Jasper credentials").Error(),
 		}
 	}
-	script, err := h.GenerateUserDataProvisioningScript(ctx, env.Settings(), creds)
+	var githubAppToken string
+	var moduleTokens []string
+	if h.ProvisionOptions != nil && h.ProvisionOptions.TaskId != "" {
+		githubAppToken, moduleTokens = units.GetGithubTokensForTask(h.ProvisionOptions.TaskId, ctx)
+	}
```
[the remark this answers is not in the document]
Comment (malikchaya2) [reply rest/data/host.go:112]: Yeah I agree that it's not ideal. But it does cause a package cycle and I don't think it's worth decoupling GenerateUserDataProvisioningScript from host struct for this, especially since we would also still need to call the function in fetchRemoteTaskData in units. 
But I'll do it if you feel strongly about it. Lmk.

File: units/provisioning_setup_host.go (hunk 23 above)
[the remark this answers is not in the document]
Comment (malikchaya2) [reply units/provisioning_setup_host.go]: (commented above)

File: operations/fetch.go [hunk 30]
```diff
@@ -127,17 +148,39 @@ func Fetch() cli.Command {
 				}
 			}
 
+			if revokeTokens {
+				err = revokeFetchTokens(ctx, client, taskID, token, moduleTokensMap)
+				grip.Warning(message.WrapError(err, message.Fields{
+					"message": "revoking github tokens after fetching data",
```
Comment (Kimchelly) [inline operations/fetch.go:154]: nit: "could not revoke GitHub tokens after fetching data"

File: operations/fetch.go [hunk 31]
```diff
@@ -127,17 +148,39 @@ func Fetch() cli.Command {
...
+			}
 			return nil
 		},
 	}
 }
 
+func parseModuleTokens(moduleTokens []string) map[string]string {
+	moduleTokensMap := make(map[string]string)
+	for _, token := range moduleTokens {
+		// parse the string formatted as 'moduleName:token'
```
Comment (Kimchelly) [inline operations/fetch.go:166]: nit: `// Parse the string`

File: rest/data/host.go (hunk 29 above)
[the remark this answers is not in the document]
Comment (Kimchelly) [reply rest/data/host.go:112]: Eh, it seems fine as it is. It's not used that much anyways.

File: rest/route/task.go [hunk 32]
```diff
@@ -383,3 +385,50 @@ func (rh *generatedTasksGetHandler) Run(ctx context.Context) gimlet.Responder {
 
 	return gimlet.NewJSONResponse(apiTaskInfos)
 }
+
+// DELETE /rest/v2/task/{task_id}/github_dynamic_access_token
+// This route is used to revoke user-used GitHub access token for a task.
```
[the remark this answers is not in the document]
Comment (Kimchelly) [reply rest/route/task.go:390]: I think this comment got missed? It looks the same.

File: rest/client/methods.go [hunk 33]
```diff
@@ -540,6 +540,25 @@ func (c *communicatorImpl) GetEstimatedGeneratedTasks(ctx context.Context, patch
...
 
+func (c *communicatorImpl) RevokeGitHubDynamicAccessTokens(ctx context.Context, taskId string, tokens []string) error {
+	info := requestInfo{
+		method: http.MethodDelete,
+		path:   fmt.Sprintf("tasks/%s/github_dynamic_access_tokens", taskId),
+	}
+
+	resp, err := c.request(ctx, info, tokens)
+	if err != nil {
+		return errors.Wrap(err, "revoking github dynamic access token")
```
Comment (Kimchelly) [inline rest/client/methods.go:551]: mega nit: `GitHub dynamic access token`

File: rest/client/methods.go [hunk 34]
```diff
@@ -540,6 +540,25 @@ func (c *communicatorImpl) GetEstimatedGeneratedTasks(ctx context.Context, patch
...
+	}
+	defer resp.Body.Close()
+
+	if resp.StatusCode != http.StatusOK {
+		return util.RespErrorf(resp, "revoking github dynamic access token")
```
Comment (Kimchelly) [inline rest/client/methods.go:556]: mega nit: `GitHub dynamic access token`

File: units/provisioning_setup_host.go (hunk 22 above)
Comment (Kimchelly) [reply units/provisioning_setup_host.go:858]: nit: could simplify this check down to `if mfest != nil`

File: units/provisioning_setup_host.go [hunk 35]
```diff
@@ -833,3 +841,59 @@ func parseLsblkOutput(lsblkOutput string) ([]blockDevice, error) {
...
+	catcher := grip.NewBasicCatcher()
+
+	var projectOwner, projectRepo string
+	var modules map[string]*manifest.Module
+	t, err := task.FindOneId(taskId)
+	catcher.Add(err)
+	if err == nil && t != nil {
+		mfest, err := manifest.FindFromVersion(t.Version, t.Project, t.Revision, t.Requester)
+		catcher.Add(err)
+		if err == nil && mfest != nil {
```
Comment (Kimchelly) [inline units/provisioning_setup_host.go:857]: nit: could simplify this check down to `if mfest != nil`

File: units/provisioning_setup_host.go [hunk 36]
```diff
@@ -833,3 +841,59 @@ func parseLsblkOutput(lsblkOutput string) ([]blockDevice, error) {
...
+			modules = mfest.Modules
+		}
+		p, err := model.FindMergedProjectRef(t.Project, t.Version, false)
+		catcher.Add(err)
+		if err == nil && p != nil {
```
Comment (Kimchelly) [inline units/provisioning_setup_host.go:862]: nit: could simplify this check down to `if p != nil`

File: units/provisioning_setup_host.go [hunk 37]
```diff
@@ -833,3 +841,59 @@ func parseLsblkOutput(lsblkOutput string) ([]blockDevice, error) {
...
+	var moduleTokens []string
+	settings, err := evergreen.GetConfig(ctx)
+	catcher.Add(err)
+	opts := &github.InstallationTokenOptions{
+		Permissions: &github.InstallationPermissions{
+			Contents: utility.ToStringPtr(thirdparty.GithubPermissionRead),
+		},
+	}
+
+	if projectOwner != "" && projectRepo != "" && err == nil {
```
Comment (Kimchelly) [inline units/provisioning_setup_host.go:877]: nit: for consistency with the rest of the function, could check `settings != nil`

File: rest/route/task.go [hunk 38]
```diff
@@ -383,3 +385,50 @@ func (rh *generatedTasksGetHandler) Run(ctx context.Context) gimlet.Responder {
...
+	catcher := grip.NewBasicCatcher()
+
+	for i, token := range h.Tokens {
+		if err := thirdparty.RevokeInstallationToken(ctx, token); err != nil {
+			catcher.Wrapf(err, "revoking token %d for task '%s'", i, h.taskID)
+		}
+	}
+
+	if catcher.HasErrors() {
+
```
Comment (Kimchelly) [inline rest/route/task.go:429]: nit: remove random whitespace.

File: rest/data/host.go [hunk 39]
```diff
@@ -104,7 +103,14 @@ func GenerateHostProvisioningScript(ctx context.Context, env evergreen.Environme
 			Message:    errors.Wrap(err, "generating Jasper credentials").Error(),
 		}
 	}
-	script, err := h.GenerateUserDataProvisioningScript(ctx, env.Settings(), creds)
+	var githubAppToken string
+	var moduleTokens []string
+	if h.ProvisionOptions != nil && h.ProvisionOptions.TaskId != "" {
+		// Do not error when trying to populate github tokens because if the repo is not private, cloning the repo will still work.
+		// Additionally, we should still spin up the host even if we can't fetch the data.
+		githubAppToken, moduleTokens, _ = units.GetGithubTokensForTask(ctx, h.ProvisionOptions.TaskId)
```
Comment (Kimchelly) [inline rest/data/host.go:111]: This needs to grip log the errors.

File: rest/data/host.go [hunk 40]
```diff
@@ -104,7 +103,14 @@ func GenerateHostProvisioningScript(ctx context.Context, env evergreen.Environme
 			Message:    errors.Wrap(err, "generating Jasper credentials").Error(),
 		}
 	}
-	script, err := h.GenerateUserDataProvisioningScript(ctx, env.Settings(), creds)
+	var githubAppToken string
+	var moduleTokens []string
+	if h.ProvisionOptions != nil && h.ProvisionOptions.TaskId != "" {
+		// Do not error when trying to populate github tokens because if the repo is not private, cloning the repo will still work.
+		// Additionally, we should still spin up the host even if we can't fetch the data.
```
Comment (Kimchelly) [inline rest/data/host.go:110]: Nice clarification comment! 🎉

File: units/provisioning_setup_host.go [hunk 41]
```diff
@@ -575,7 +580,10 @@ func (j *setupHostJob) fetchRemoteTaskData(ctx context.Context) error {
 	if j.host.ProvisionOptions.TaskSync {
 		cmd = strings.Join(j.host.SpawnHostPullTaskSyncCommand(), " ")
 	} else {
-		cmd = strings.Join(j.host.SpawnHostGetTaskDataCommand(), " ")
+		// Do not error when trying to populate github tokens because if the repo is not private, cloning the repo will still work.
+		// Additionally, we should still spin up the host even if we can't fetch the data
+		githubAppToken, moduleTokens, _ := GetGithubTokensForTask(ctx, j.host.ProvisionOptions.TaskId)
```
Comment (Kimchelly) [inline units/provisioning_setup_host.go:585]: This needs to grip log the errors.

**Review by Kimchelly — APPROVED** [review]
LGTM mod some non-blocking nits. Nice work!
